### ISSUE TITLE
[vm/ffi] Updating double-conversion causes `std::cout` to segfault in FFI calls on MacOS

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#47835]
Updating double-conversion causes `std::cout` to segfault in ffi calls.

* https://dart-review.googlesource.com/c/sdk/+/221884
* https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8829006337907998913/+/u/test_results/new_test_failures__logs_

Minimal repro (any FFI call to native code with `std::cout`).

```dart
import 'dart:ffi';

import 'ffi_test_helpers.dart';

void main() {
  final takeMaxUint8 = ffiTestFunctions
      .lookupFunction<IntPtr Function(Int8), int Function(int)>("TakeMaxUint8");
  takeMaxUint8(0xFF);
}
```

Initial investigation:

* The FFI `Trampoline`s are identical in IL and machine code.
* Setting a breakpoint on `TakeMaxUint8` gives a very similar looking registers and stack. So we don't have a misaligned stack or wrong stack height.
* Using the the `libffi_test_functions.dylib` from the other (top of tree vs CL) does not make a difference. So the problem should be in the Dart VM, not in the dylib.

Earlier encounters with `std::cout` segfaults

* https://github.com/dart-lang/sdk/issues/43075
* misaligned stack, causing compiled C++ code which is supposed to refer to a pointer read garbage and try to read from wrong memory address (which is my hunch on what is happening here, but I don't see it.)

Blocking:

* https://github.com/dart-lang/sdk/issues/38587

cc @rmacnak-google

**Comment by dcharkes** [comment]
I investigated the top two frames of the segfault and their machine code.

```
dart!std::__2::collate<wchar_t>::do_hash(wchar_t const*, wchar_t const*) const
dart!std::__2::__num_put<char>::__widen_and_group_int(char*, char*, char*, char*, char*&, char*&, std::__2::locale const&)
```

`do_hash` loops from the first to the second pointer. We segfault because `0x1` is passed in to the second argument, causing an infinite loop until we cross the memory page boundary causing invalid access.

`__widen_and_group_int` computes the argument with
```
    0x100019747: 48 63 05 9a c9 3f 02     movslq 0x23fc99a(%rip), %rax     ; std::__2::numpunct<char>::id + 8
    0x10001974e: 48 8d 50 ff              leaq   -0x1(%rax), %rdx
```
where `id + 8` is `0x2` so the argument becomes `0x1`. That also seems reasonable.

What does not seem reasonable is the call instruction leading to `do_hash`. When I open the source code in `buildtools/mac-x64/clang/include/c++/v1/locale` there is no call to `do_hash` in `__widen_and_group_int`.

I wonder if we have two different versions of libc++ linked somehow.

Then it would be likely that that is also the cause of `std::cout` issues in

* https://github.com/dart-lang/sdk/issues/43075

**Comment by dcharkes** [comment]
This might actually be a bug in our toolchain.

Compiling a dynamic library which invokes std::cout in a separate CMake project on MacOS works fine when using XCode's clang version:

```
$ /Applications/Xcode13.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ --version
Apple clang version 13.0.0 (clang-1300.0.29.3)
```

Adding the following the flags to the CMake invocation causes exactly the error that we're seeing:

```
-DCMAKE_CXX_COMPILER=/Users/dacoharkes/dart-sdk/sdk/buildtools/mac-x64/clang/bin/clang++
-DCMAKE_CXX_FLAGS="-Wl,/Users/dacoharkes/dart-sdk/sdk/buildtools/mac-x64/clang/lib/libc++.a"
```

Only adding the first flag makes the compilation fail with `std::__2` symbols missing:

```
Undefined symbols for architecture x86_64:
  "std::__2::locale::use_facet(std::__2::locale::id&) const", referenced from:
      std::__2::ctype<char> const& std::__2::use_facet<std::__2::ctype<char> >(std::__2::locale const&) in mylib_dylib.cc.o
...
```

We started passing in a `libc++.a` manually in 73f6d15665a3c9a7cb521c90f7257c606c53171a.

Last clang roll was in ba0b0d8cc6d497cc1963c381f6ffcbc457b89c5e.

```
$ buildtools/mac-x64/clang/bin/clang++ --version
Fuchsia clang version 14.0.0 (https://llvm.googlesource.com/a/llvm-project e3a7f0e2f9ab566bd9b71fb54fe77e947b061a12)
```

@rmacnak-google why do we have to pass in `lib/libc++.a` in our build?

**Comment by rmacnak-google** [comment]
I'm not sure why we're explicitly passing `libc++.a`. IIRC, this also created some problems with the sanitizers, which disable it. I'll try building with the default.

**Comment by dcharkes** [comment]
Another thing I tried but didn't document yet: don't link in the standard library to the dynamic library and resolve symbols at runtime with `-undefined dynamic_lookup`. However, that makes lookup of std symbols fail when trying to load the dynamic library.

```
Unhandled exception:
Invalid argument(s): Failed to load dynamic library '/Users/dacoharkes/dart-sdk/sdk/xcodebuild/DebugX64/libffi_test_functions.dylib': dlopen(/Users/dacoharkes/dart-sdk/sdk/xcodebuild/DebugX64/libffi_test_functions.dylib, 0x0001): symbol not found in flat namespace '__ZNSt3__217bad_function_callD1Ev'
#0      _open (dart:ffi-patch/ffi_dynamic_library_patch.dart:12:43)
#1      new DynamicLibrary.open (dart:ffi-patch/ffi_dynamic_library_patch.dart:23:12)
#2      main (file:///Users/dacoharkes/dart-sdk/sdk/tests/ffi/aliasing_test.dart:14:32)
#3      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#4      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
```

**Comment by cheatcodezky** [comment]
已收到您的邮件，我会尽快回复。

**Comment by rmacnak-google** [comment]
If I try to use the default standard library, I get missing symbols at link time. Maybe the Clang we get from Fuchsia is using different headers by default?

**Comment by rmacnak-google** [comment]
I think this was fixed by b8ee3a999676b0e502c0a2a9cba9730374d9bf83.

New roll attempt in https://dart-review.googlesource.com/c/sdk/+/325543.
